pkg/operator/nfs/controller.go

```diff
@@ -0,0 +1,422 @@
+	spec      nfsv1alpha1.NFSServerSpec
+	ownerRef  metav1.OwnerReference
+}
+
+func newNfsServer(c *nfsv1alpha1.NFSServer, context *clusterd.Context) *nfsServer {
+	return &nfsServer{
+		name:      appName,
+		context:   context,
+		namespace: c.Namespace,
+		spec:      c.Spec,
+		ownerRef:  nfsOwnerRef(c.Namespace, string(c.UID)),
+	}
+}
+
+func nfsOwnerRef(namespace, nfsServerID string) metav1.OwnerReference {
+	blockOwner := true
+	return metav1.OwnerReference{
+		APIVersion:         NFSResource.Version,
+		Kind:               NFSResource.Kind,
+		Name:               namespace,
+		UID:                types.UID(nfsServerID),
+		BlockOwnerDeletion: &blockOwner,
+	}
+}
+
+func getServerConfig(exports []nfsv1alpha1.ExportsSpec) map[string]map[string]string {
+	claimConfigOpt := make(map[string]map[string]string)
+	configOpt := make(map[string]string)
+
+	for _, export := range exports {
+		claimName := export.PersistentVolumeClaim.ClaimName
+		if claimName != "" {
+			configOpt["accessMode"] = export.Server.AccessMode
+			configOpt["squash"] = export.Server.Squash
+			claimConfigOpt[claimName] = configOpt
+		}
+	}
+
+	return claimConfigOpt
+}
+
+func createAppLabels() map[string]string {
+	return map[string]string{
+		k8sutil.AppAttr: appName,
+	}
+}
+
+func createServicePorts() []v1.ServicePort {
+	return []v1.ServicePort{
+		{
+			Name:       "nfs",
+			Port:       int32(nfsPort),
+			TargetPort: intstr.FromInt(int(nfsPort)),
+		},
+		{
+			Name:       "rpc",
+			Port:       int32(rpcPort),
+			TargetPort: intstr.FromInt(int(rpcPort)),
+		},
+	}
+}
+
+func (c *Controller) createNFSService(nfsServer *nfsServer) error {
+	// This service is meant to be used by clients to access NFS.
+	nfsService := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            nfsServer.name,
+			Namespace:       nfsServer.namespace,
+			OwnerReferences: []metav1.OwnerReference{nfsServer.ownerRef},
+			Labels:          createAppLabels(),
+		},
+		Spec: v1.ServiceSpec{
+			Selector: createAppLabels(),
+			Type:     v1.ServiceTypeClusterIP,
+			Ports:    createServicePorts(),
+		},
+	}
+
+	if _, err := c.context.Clientset.CoreV1().Services(nfsServer.namespace).Create(nfsService); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("nfs service %s already exists in namespace %s", nfsService.Name, nfsService.Namespace)
+	} else {
+		logger.Infof("nfs service %s started in namespace %s", nfsService.Name, nfsService.Namespace)
+	}
+
+	return nil
+}
+
+func createGaneshaExport(id int, path string, access string, squash string) string {
+	var accessType string
+	if access == "ReadWrite" {
+		accessType = "RW"
+	}
+	idStr := fmt.Sprintf("%v", id)
+	nfsGaneshaConfig := `
+EXPORT {
+	Export_Id = ` + idStr + `;
+	Path = /` + path + `;
+	Pseudo = /` + path + `;
+	Protocols = 4;
+	Transports = TCP;
+	Sectype = sys;
+	Access_Type = ` + accessType + `;
+	Squash = ` + squash + `;
+	FSAL {
+		Name = VFS;
+	}
+}
+NFS_Core_Param
```

Comment:
this function is called in a for loop, so this top level `NFS_Core_Param` block will be duplicated multiple times in the generated config file.  That won't work when there are more than 1 export (which we have documentation for).